common/oauthTokenManager.go

```diff
@@ -148,309 +131,92 @@ func (uotm *UserOAuthTokenManager) GetTokenInfo(ctx context.Context) (*OAuthToke
-	if err := identityInfo.Validate(); err != nil {
-		return nil, err
+func (uotm *UserOAuthTokenManager) validateAndPersistLogin(oAuthTokenInfo *OAuthTokenInfo, persist bool) error {
+	// Use default tenant ID and active directory endpoint, if nothing specified.
+	if oAuthTokenInfo.Tenant == "" {
+		oAuthTokenInfo.Tenant = DefaultTenantID
 	}
-
-	oAuthTokenInfo := &OAuthTokenInfo{
-		Identity:     true,
-		IdentityInfo: identityInfo,
+	if oAuthTokenInfo.ActiveDirectoryEndpoint == "" {
+		oAuthTokenInfo.ActiveDirectoryEndpoint = DefaultActiveDirectoryEndpoint
 	}
-	token, err := oAuthTokenInfo.GetNewTokenFromMSI(ctx)
+	tc, err := oAuthTokenInfo.GetTokenCredential()
 	if err != nil {
-		return nil, err
+		return err
+	}
+	scopes := []string{StorageScope}
+	_, err = tc.GetToken(context.TODO(), policy.TokenRequestOptions{Scopes: scopes})
+	if err != nil {
+		return err
 	}
-	oAuthTokenInfo.Token = *token
 	uotm.stashedInfo = oAuthTokenInfo
 
-	if persist {
+	if persist && err == nil {
 		err = uotm.credCache.SaveToken(*oAuthTokenInfo)
 		if err != nil {
-			return nil, err
+			return err
 		}
 	}
 
-	return oAuthTokenInfo, nil
+	return nil
 }
 
-// secretLoginNoUOTM non-interactively logs in with a client secret.
-func secretLoginNoUOTM(tenantID, activeDirectoryEndpoint, secret, applicationID, resource string) (*OAuthTokenInfo, error) {
-	if tenantID == "" {
-		tenantID = DefaultTenantID
-	}
-
-	if activeDirectoryEndpoint == "" {
-		activeDirectoryEndpoint = DefaultActiveDirectoryEndpoint
-	}
-
-	if applicationID == "" {
-		return nil, fmt.Errorf("please supply your OWN application ID")
-	}
-
-	oAuthTokenInfo := OAuthTokenInfo{
-		Tenant:                  tenantID,
-		ActiveDirectoryEndpoint: activeDirectoryEndpoint,
-	}
-
-	oauthConfig, err := adal.NewOAuthConfig(activeDirectoryEndpoint, tenantID)
-	if err != nil {
-		return nil, err
-	}
-
-	spt, err := adal.NewServicePrincipalToken(
-		*oauthConfig,
-		applicationID,
-		secret,
-		resource,
-	)
-	if err != nil {
-		return nil, err
-	}
-
-	err = spt.Refresh()
-	if err != nil {
-		return nil, err
+// MSILogin tries to get token from MSI, persist indicates whether to cache the token on local disk.
+func (uotm *UserOAuthTokenManager) MSILogin(identityInfo IdentityInfo, persist bool) error {
+	if err := identityInfo.Validate(); err != nil {
+		return err
 	}
 
-	// Due to the nature of SPA, no refresh token is given.
-	// Thus, no refresh token is copied or needed.
-	oAuthTokenInfo.Token = spt.Token()
-	oAuthTokenInfo.ApplicationID = applicationID
-	oAuthTokenInfo.ServicePrincipalName = true
-	oAuthTokenInfo.SPNInfo = SPNInfo{
-		Secret:   secret,
-		CertPath: "",
+	oAuthTokenInfo := &OAuthTokenInfo{
+		Identity:     true,
+		IdentityInfo: identityInfo,
 	}
 
-	return &oAuthTokenInfo, nil
+	return uotm.validateAndPersistLogin(oAuthTokenInfo, persist)
 }
 
 // SecretLogin is a UOTM shell for secretLoginNoUOTM.
```

Comment:
This comment isn't entirely true anymore.

Reply:
Fixed the comment.